domino/lib/src/experimental/idom_server.dart

```diff
@@ -0,0 +1,250 @@
+  _IdomElem get _shdElement => _shdPath.last;
+  @override
+  _IdomNode get pointer => _index.last < _shdElement.nodes.length
+      ? _shdElement.nodes[_index.last]
+      : null;
+
+  ServerDomContext([_IdomElem root]) : _rootElem = root ?? _IdomElem(null) {
+    _index.add(0);
+    _path.add(_rootElem);
+    _shdPath.add(_rootElem);
+  }
+
+  // TODO: onCreate, onRemove?
+  @override
+  void open(String tag, {String key, onCreate, onRemove}) {
+    // Create a new pseudo-element with empty properties
+    final newElem = _IdomElem(tag, key: key, parent: element);
+
+    // Pull nodes list if matches an element from the list.
+    final match = _shdElement.nodes.indexWhere(
+        (node) =>
+            (node is _IdomElem) &&
+            (node.tag == tag) &&
+            (key == null || key == node.key),
+        _index.last);
+
+    if (match == -1) {
+      // no match, insert new elem at the current index
+      _shdElement.nodes.insert(_index.last, newElem);
+      _index.last = _index.last + 1;
+      _path.add(newElem);
+      _shdPath.add(newElem);
+      _index.add(0);
+    } else {
+      // match, remove everything between the index and the match
+      _shdElement.nodes.removeRange(_index.last, match);
+      _path.add(pointer as _IdomElem);
+      _shdPath.add(newElem);
+      _index.add(0);
+    }
+  }
+
+  @override
+  void text(String value) {
+    final ptr = pointer;
+    if (ptr is _IdomText) {
+      // Next node is a text
+      if (ptr.text != value) {
+        ptr.text = value;
+      }
+    } else {
+      // Insert text node
+      final newText = _IdomText(value, element);
+      _shdElement.nodes.insert(_index.last, newText);
+      _index.last = _index.last + 1;
+    }
+  }
+
+  @override
+  void close({String tag}) {
+    // Remove unwalked nodes
+    _shdElement.nodes.removeRange(_index.last, _shdElement.nodes.length);
+
+    // Deep copy
+    _path.last.moveFrom(_shdPath.last);
+    _path.removeLast();
+    _shdPath.removeLast();
+    _index.removeLast();
+  }
+
+  @override
+  void attr(String name, String value) {
+    _shdElement.attr[name] = value;
+  }
+
+  @override
+  void clazz(String name, {bool present = true}) {
+    if (present) {
+      _shdElement.clazz.add(name);
+    } else {
+      _shdElement.clazz.remove(name);
+    }
+  }
+
+  @override
+  void style(String name, String value) {
+    _shdElement.style[name] = value;
+  }
+
+  @override
+  void innerHtml(String value) {
+    _shdElement.nodes = [_IdomHtml(value, element)];
+    _index.last = 1;
+  }
+
+  @override
+  void skipNode() {
+    if (_index.last < element.nodes.length) {
+      _index.last = _index.last + 1;
+    }
+  }
+
+  @override
+  void skipRemainingNodes() {
+    _index.last = element.nodes.length;
+  }
+
+  @override
+  void event(String name, {fn, String key, bool tracked = true}) {
+    // TODO: implement event
+  }
+
+  void writeHTML(StringSink out,
+      {_IdomElem elem,
+      String indent,
```

Comment:
- `indent` should be a top-level attribute of the context
- `indEnd` should be pre-calculated